diff/windows/windows.go

```diff
@@ -158,7 +164,136 @@ func (s windowsDiff) Apply(ctx context.Context, desc ocispec.Descriptor, mounts
 // Compare creates a diff between the given mounts and uploads the result
 // to the content store.
 func (s windowsDiff) Compare(ctx context.Context, lower, upper []mount.Mount, opts ...diff.Opt) (d ocispec.Descriptor, err error) {
-	return emptyDesc, errors.Wrap(errdefs.ErrNotImplemented, "windowsDiff does not implement Compare method")
+	t1 := time.Now()
+
+	var config diff.Config
+	for _, opt := range opts {
+		if err := opt(&config); err != nil {
+			return emptyDesc, err
+		}
+	}
+
+	layers, err := mountPairToLayerStack(lower, upper)
+	if err != nil {
+		return emptyDesc, err
+	}
+
+	if config.MediaType == "" {
+		config.MediaType = ocispec.MediaTypeImageLayerGzip
+	}
+
+	var isCompressed bool
+	switch config.MediaType {
+	case ocispec.MediaTypeImageLayer:
+	case ocispec.MediaTypeImageLayerGzip:
+		isCompressed = true
+	default:
+		return emptyDesc, errors.Wrapf(errdefs.ErrNotImplemented, "unsupported diff media type: %v", config.MediaType)
```

Comment:
Can we support Docker media types?

Reply:
Ummm. Weird. The [walking difffer](https://github.com/containerd/containerd/blob/dd14169771817b2674c22240ca63ba54bdd0f803/diff/walking/differ.go#L68-L79) doesn't list docker types here either.

Is it possible that a different layer has already folded the Docker media types into their OCI equivalents, i.e. are we passing through, e.g., [here](https://github.com/containerd/containerd/blob/a1138182d5d746351b3c321893001e33fd7fb4be/images/converter/default.go#L79-L93) or similar first, for all the code-paths that arrive here?

Otherwise, I think that might be a different change, that could be applied to both existing `Compare` implementations at once. The third differ, LCOW, doesn't have an implementation of `Compare` yet.

Reply:
Following up on this, I did notice that `TestImport` covers [some Docker media types](https://github.com/containerd/containerd/blob/85041ffe58751b27e3fc068a1d605e058d12e4aa/integration/client/import_test.go#L143-L176), but `TestExportAndImport` and `TestExport` do not, as it appears that the Export format always produces OCI layers per the OCI image spec, and then adds in a Docker-compatible manifest.

In fact, I can't see any code-path except `ctr snapshot diff --media-type` that could get a media type setting into this API, which is a _huge_ hole in the test suite. We do expose `diff.WithMediaType` for use with `client.DiffService().Compare` and a `MediaType` field in `api/services/diff/v1.DiffRequest` for API users who might reasonably expect to be able to get Docker layer types out, since they were able to *import* Docker media types, but we never test those options, that I can see. >_<

Reply:
I created #5317 so this isn't forgotten.